server/src/main/java/org/opensearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -288,13 +296,132 @@ private static class FromSingle extends AutoDateHistogramAggregator {
+                }
+            }
+        }
+
+        private void findBoundsAndCreateFilters(final String fieldName, final SearchContext context) throws IOException {
+            final List<LeafReaderContext> leaves = context.searcher().getIndexReader().leaves();
+            long min = Long.MAX_VALUE, max = Long.MIN_VALUE;
+            // Since the query does not specify bounds for aggregation, we can
+            // build the global min/max from local min/max within each segment
+            for (LeafReaderContext leaf : leaves) {
+                min = Math.min(min, NumericUtils.sortableBytesToLong(
+                    leaf.reader().getPointValues(fieldName).getMinPackedValue(), 0));
+                max = Math.max(max, NumericUtils.sortableBytesToLong(
+                    leaf.reader().getPointValues(fieldName).getMaxPackedValue(), 0));
+            }
+            createFilterForAggregations(fieldName, min, max);
+        }
+
+        private void createFilterForAggregations(final String field, final long low, final long high) throws IOException {
+            // max - min / targetBuckets = bestDuration
+            // find the right innerInterval this bestDuration belongs to
+            // since we cannot exceed targetBuckets, bestDuration should go up,
+            // so the right innerInterval should be an upper bound
+            long bestDuration = (high - low) / targetBuckets;
+            long interval = 0;
+            roundingInfosLoop: do {
+                RoundingInfo curRoundingInfo = roundingInfos[roundingIdx];
+                for (int curInnerInterval: curRoundingInfo.innerIntervals) {
+                    if (bestDuration <= curInnerInterval * curRoundingInfo.roughEstimateDurationMillis) {
+                        interval = curInnerInterval * curRoundingInfo.roughEstimateDurationMillis;
+                        break roundingInfosLoop;
+                    }
+                }
+                roundingIdx++;
+            } while (
+                roundingIdx < roundingInfos.length
+            );
+
+            // interval > 0
+            int i = 0;
+            filters = new Weight[targetBuckets];
+            long roundedLow = preparedRounding.round(low);
+            while (i < targetBuckets) {
+                // Calculate the lower bucket bound
+                final byte[] lower = new byte[8];
+                NumericUtils.longToSortableBytes(Math.max(roundedLow, low), lower, 0);
+                // Calculate the upper bucket bound
+                final byte[] upper = new byte[8];
+                roundedLow = preparedRounding.round(roundedLow + interval);
+                // Subtract -1 if the minimum is roundedLow as roundedLow itself
+                // is included in the next bucket
+                NumericUtils.longToSortableBytes(Math.min(roundedLow - 1, high), upper, 0);
+
+                filters[i++] = context.searcher().createWeight(new PointRangeQuery(field, lower, upper, 1) {
```

Comment:
This logic has problem since we are focused on creating the required number of target buckets, instead we should limit to the upper and lower bounds. For example, I see the following response:

```
% curl -s -X GET "localhost:9200/nyc_taxis/_search?pretty" -H 'Content-Type: application/json' -d'{"size": 0,"query": {"range": {"dropoff_datetime": {"gte": "2015-01-01 01:04:06","lt": "2016-01-01 00:00:00"}}},"aggs": {"dropoffs_over_time": {"auto_date_histogram": {"field": "dropoff_datetime","buckets": "4"}}}}'
{
  "took" : 4295,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 100,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "dropoffs_over_time" : {
      "buckets" : [
        {
          "key_as_string" : "2015-01-01 00:00:00",
          "key" : 1420070400000,
          "doc_count" : 100
        },
        {
          "key_as_string" : "2016-01-01 00:00:00",
          "key" : 1451606400000,
          "doc_count" : 0
        },
        {
          "key_as_string" : "2017-01-01 00:00:00",
          "key" : 1483228800000,
          "doc_count" : 0
        }
      ],
      "interval" : "1y"
    }
  }
}
```